include/signalrclient/signalr_value.h

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#pragma once
+
+#include <string>
+#include <vector>
+#include <map>
+
+namespace signalr
+{
+    /**
+     * An enum defining the types a signalr::value may be.
+     */
+    enum class type
+    {
+        MAP,
+        ARRAY,
+        STRING,
+        NUMBER,
+        NULL,
+        BOOL
+    };
+
+    /**
+     * Represents a value to be provided to a SignalR method as a parameter, or returned as a return value.
+     */
+    class value
+    {
+    public:
+        value();
+        value(int val);
+        value(double val);
+        value(const std::string& val);
+        value(const std::vector<value>& val);
+        value(const std::map<std::string, value>& map);
+
+        /**
+         * True if the object stored is a Key-Value pair.
+         */
+        bool is_map() const;
+
+        /**
+         * True if the object stored is a number (int, float, char).
+         */
+        bool is_number() const;
+
+        /**
+         * True if the object stored is a string.
+         */
+        bool is_string() const;
+
+        /**
+         * True if the object stored is empty.
+         */
+        bool is_null() const;
+
+        /**
+         * True if the object stored is an array of signalr::value's.
+         */
+        bool is_array() const;
+
+        /**
+         * True if the object stored is a bool.
+         */
+        bool is_bool() const;
+
+        /**
+         * Returns the stored object as a double. The value will be cast to a double if the underlying object is a signalr::type::NUMBER otherwise it will throw
+         */
+        double as_double() const;
+
+        /**
+         * Returns the stored object as a float. The value will be cast to a float if the underlying object is a signalr::type::NUMBER otherwise it will throw.
+         */
+        float as_float() const;
+
+        /**
+         * Returns the stored object as an int. The value will be cast to an int if the underlying object is a signalr::type::NUMBER otherwise it will throw.
+         */
+        int as_int() const;
```

Comment:
I'm not sure that always casting a long/float/double to an int is the best thing here. I think we should

1. throw from as_int() if the value isn't an int or directly translatable without rounding/truncating/clamping
2. provide an is_int(), is_double(), is_long() etc, so you don't have to call as_int() *and* as_double() and verify they're the same in order to verify you *actually* got an int.

Reply:
It's probably even better is only providing one way to get a number (probably as double) and letting the caller cast to an int if they really want to.

Reply:
> It's probably even better to only providing one way to get a number (probably as double) and letting the caller cast to an int if they really want to

One reason for having all the different types is to support things like MessagePack in the future, where int and long are actual types in the spec and so we would add new types to the enum and for example throw in as_int if the type is not NUMBER or INT.

I could be convinced to only have double for now and add the others later as needed.

Reply:
> I could be convinced to only have double for now and add the others later as needed.

Let's do that.

Reply:
Yeah, I was kinda feeling that way but not enough to say so :P. JSON "numbers" are always floating point. I'd even go as far as saying we rename `NUMBER` to `DOUBLE` to give us flexibility later on with MsgPack.

Reply:
I actually like that, NUMBER is very JSON specific. Part of the point of this is to be a generic object model. So JSON will just always put numbers in a double.